15-chatroom/src/entry.py

```diff
@@ -0,0 +1,136 @@
+from workers import WorkerEntrypoint, Response, DurableObject
+from pathlib import Path
+from js import WebSocketPair, Object
+import json
+from urllib.parse import urlparse
+from datetime import datetime, timezone
+
+
+class Chatroom(DurableObject):
+    """Durable Object that manages a chatroom with WebSocket connections."""
+
+    def __init__(self, state, env):
+        super().__init__(state, env)
+        self.state = state
+        self.env = env
+        self.message_history = []  # Limited message history
+        self.max_history = 50  # Maximum number of messages to keep
+
+    async def fetch(self, request):
+        """Handle incoming requests to the Durable Object."""
+
+        # Check if this is a WebSocket upgrade request
+        upgrade_header = request.headers.get("Upgrade")
+        if upgrade_header and upgrade_header.lower() == "websocket":
+            # Create a WebSocket pair
+            pair = Object.values(WebSocketPair.new())
+            client = pair[0]
+            server = pair[1]
+
+            # Accept the WebSocket connection - this tells the DO to handle it
+            self.state.acceptWebSocket(server)
+
+            # Send message history to the newly connected client
+            if self.message_history:
+                history_msg = {"type": "history", "messages": self.message_history}
+                server.send(json.dumps(history_msg))
+
+            # Send a welcome message
+            welcome_msg = {
+                "type": "system",
+                "text": "Connected to chatroom",
+                "timestamp": self.get_timestamp(),
+            }
+            server.send(json.dumps(welcome_msg))
+
+            # Return the client-side WebSocket in the response
+            return Response(None, status=101, web_socket=client)
+
+        # If not a WebSocket request, return an error
+        return Response("Expected WebSocket upgrade", status=400)
```

Comment:
Can we handle the error case first and then move the main case out of the conditional block?